src/map_votes.cpp

```diff
@@ -0,0 +1,543 @@
+		int iNumNominations = g_pMapVoteSystem->GetTotalNominations(iResponse);
+		ClientPrintAll(HUD_PRINTTALK, CHAT_PREFIX " \x06%s \x01was nominated by %s. It now has %d nominations.", sMapName, sPlayerName, iNumNominations);
+	}
+}
+
+static int __cdecl OrderStringsLexicographically(const char* const* a, const char* const* b)
+{
+	return V_strcasecmp(*a, *b);
+}
+
+CON_COMMAND_CHAT_FLAGS(maplist, "List the maps in the server", ADMFLAG_NONE)
+{
+	ClientPrint(player, HUD_PRINTTALK, CHAT_PREFIX "The list of all maps will be shown in console.");
+	ClientPrint(player, HUD_PRINTCONSOLE, "The list of all maps is:");
+	CUtlVector<const char*> vecMapNames;
+	for (int i = 0; i < g_pMapVoteSystem->GetMapListSize(); i++) {
+		vecMapNames.AddToTail(g_pMapVoteSystem->GetMapName(i));
+	}
+	vecMapNames.Sort(OrderStringsLexicographically);
+	FOR_EACH_VEC(vecMapNames, i)
+	{
+		ClientPrint(player, HUD_PRINTCONSOLE, "- %s", vecMapNames[i]);
+	}
+}
+
+CON_COMMAND_CHAT_FLAGS(nomlist, "List the list of nominations", ADMFLAG_NONE)
+{
+	ClientPrint(player, HUD_PRINTTALK, CHAT_PREFIX "Current nominations:");
+	for (int i = 0; i < g_pMapVoteSystem->GetMapListSize(); i++) {
+		if (!g_pMapVoteSystem->IsMapIndexEnabled(i)) continue;
+		int iNumNominations = g_pMapVoteSystem->GetTotalNominations(i);
+		if (iNumNominations == 0) continue;
+		const char* sMapName = g_pMapVoteSystem->GetMapName(i);
+		ClientPrint(player, HUD_PRINTTALK, CHAT_PREFIX "- %s (%d times)\n", sMapName, iNumNominations);
+	}
+}
+
+CON_COMMAND_CHAT_FLAGS(mapcooldowns, "List the maps currently in cooldown", ADMFLAG_NONE)
+{
+	ClientPrint(player, HUD_PRINTTALK, CHAT_PREFIX "The list of maps in cooldown will be shown in console.");
+	ClientPrint(player, HUD_PRINTCONSOLE, "The list of maps in cooldown is:");
+	int iMapsInCooldown = g_pMapVoteSystem->GetMapsInCooldown();
+	for (int i = iMapsInCooldown - 1; i >= 0; i--) {
+		int iMapIndex = g_pMapVoteSystem->GetCooldownMap(i);
+		const char* sMapName = g_pMapVoteSystem->GetMapName(iMapIndex);
+		ClientPrint(player, HUD_PRINTCONSOLE, "- %s (%d maps ago)", sMapName, iMapsInCooldown - i);
+	}
+}
+
+GAME_EVENT_F(cs_win_panel_match)
+{
+	g_pMapVoteSystem->StartVote();
+}
+
+GAME_EVENT_F(endmatch_mapvote_selecting_map)
+{
+	g_pMapVoteSystem->FinishVote();
+}
+
+bool CMapVoteSystem::IsMapIndexEnabled(int iMapIndex)
+{
+	if (iMapIndex >= m_vecMapList.Count() || iMapIndex < 0) return false;
+	if (m_vecLastPlayedMapIndexes.HasElement(iMapIndex)) return false;
+	return m_vecMapList[iMapIndex].IsEnabled();
+}
+
+void CMapVoteSystem::StartVote() 
+{
+	m_bIsVoteOngoing = true;
+
+	// If we are forcing a map, just set all vote options to that map
+	if (m_iForcedNextMapIndex != -1) {
+		for (int i = 0; i < 10; i++) {
+			g_pGameRules->m_nEndMatchMapGroupVoteOptions[i] = m_iForcedNextMapIndex;
+			return;
+		}
+	}
+
+	// Seed the randomness for the event
+	int iRandomSeed = (int)gpGlobals->curtime;
+	srand(iRandomSeed);
```

Comment:
We already `srand` once when the plugin is initialized, there's no need to do it again.

Also if you're only using `m_iRandomWinnerShift` once elsewhere, why generate the random number here. And for that matter, why even have it as a member in the first place?

Reply:
We generate a single random shift per vote, such that as players vote (and create new ties), the way to decide how to tie-break remains deterministic. I'll remove srand though.